lib/winpki/ldap.go

```diff
@@ -283,3 +309,68 @@ func crlKeyName(caType types.CertAuthType) string {
 		return "Teleport"
 	}
 }
+
+// createConnection dials an LDAP server using the provided TLS config.
+// The server is either obtained directly from the configuration or
+// discovered via DNS.
+func (c *LDAPConfig) createConnection(ctx context.Context, ldapTLSConfig *tls.Config) (*ldap.Conn, error) {
+	servers := []string{c.Addr}
+	dialer := net.Dialer{Timeout: ldapDialTimeout}
+
+	if c.LocateServer.Enabled {
+		dial := func(dialCtx context.Context, network, address string) (net.Conn, error) {
+			return dialer.DialContext(dialCtx, network, address)
+		}
+
+		// In development environments, the system's default resolver is unlikely to be
+		// able to resolve the Active Directory SRV records needed for server location,
+		// so we allow overriding the resolver.
+		if resolverAddr := os.Getenv("TELEPORT_LDAP_RESOLVER"); resolverAddr != "" {
+			c.Logger.DebugContext(ctx, "Using custom DNS resolver address", "address", resolverAddr)
+			// Check if resolver address has a port
+			host, port, err := net.SplitHostPort(resolverAddr)
+			if err != nil {
+				host = resolverAddr
+				port = "53"
+			}
+
+			customResolverAddr := net.JoinHostPort(host, port)
+			dial = func(ctx context.Context, network, address string) (net.Conn, error) {
+				return dialer.DialContext(ctx, network, customResolverAddr)
+			}
+		}
+
+		resolver := &net.Resolver{
+			PreferGo: true,
+			Dial:     dial,
+		}
+
+		var err error
+		if servers, err = locateLDAPServer(ctx, c.Domain, c.LocateServer.Site, resolver); err != nil {
+			return nil, trace.Wrap(err, "locating LDAP server")
+		}
+	}
+
+	if len(servers) == 0 {
+		return nil, trace.NotFound("no LDAP servers found for domain %q", c.Domain)
+	}
+
+	for _, server := range servers {
+		conn, err := ldap.DialURL(
+			"ldaps://"+server,
+			ldap.DialWithDialer(&dialer),
+			ldap.DialWithTLSConfig(ldapTLSConfig),
+		)
+
+		if err != nil {
+			// If the connection fails, try the next server
+			c.Logger.DebugContext(ctx, "Error connecting to LDAP server, trying next available server", "server", server, "error", err)
+			continue
+		}
+
```

Comment:
Let's emit a debug log here so that we can see what the result of the server location was when we read the logs.